SignalServiceKit/src/Storage/OWSStorage.m

```diff
@@ -0,0 +1,505 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSStorage.h"
+#import "AppContext.h"
+#import "NSData+Base64.h"
+#import "TSAttachmentStream.h"
+#import "TSStorageManager.h"
+#import <Curve25519Kit/Randomness.h>
+#import <SAMKeychain/SAMKeychain.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString *const OWSStorageExceptionName_DatabasePasswordInaccessibleWhileBackgrounded
+    = @"OWSStorageExceptionName_DatabasePasswordInaccessibleWhileBackgrounded";
+NSString *const OWSStorageExceptionName_DatabasePasswordUnwritable
+    = @"OWSStorageExceptionName_DatabasePasswordUnwritable";
+NSString *const OWSStorageExceptionName_NoDatabase = @"OWSStorageExceptionName_NoDatabase";
+
+static NSString *keychainService = @"TSKeyChainService";
+static NSString *keychainDBPassAccount = @"TSDatabasePass";
+
+#pragma mark -
+
+@interface YapDatabaseConnection ()
+
+- (id)initWithDatabase:(YapDatabase *)database;
+
+@end
+
+#pragma mark -
+
+@interface OWSDatabaseConnection : YapDatabaseConnection
+
+@property (atomic, weak) id<OWSDatabaseConnectionDelegate> delegate;
+
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithDatabase:(YapDatabase *)database
+                        delegate:(id<OWSDatabaseConnectionDelegate>)delegate NS_DESIGNATED_INITIALIZER;
+
+@end
+
+#pragma mark -
+
+@implementation OWSDatabaseConnection
+
+- (id)initWithDatabase:(YapDatabase *)database delegate:(id<OWSDatabaseConnectionDelegate>)delegate
+{
+    self = [super initWithDatabase:database];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(delegate);
+
+    _delegate = delegate;
+
+    return self;
+}
+
+// This clobbers the superclass implementation to include an assert which
+// ensures that the database is in a ready state before creating write transactions.
+//
+// Creating write transactions before the _sync_ database views are registered
+// causes YapDatabase to rebuild all of our database views, which is catastrophic.
+// We're not sure why, but it causes YDB's "view version" checks to fail.
```

Comment:
In a forthcoming PR we'll do more work in these custom database/connection classes - around background tasks, for example.

Reply:
That seems pretty reasonable.  This is a legacy comment - I'll update it a bit here, but these methods are going to get reworked in the coming days.